extensions/panache/hibernate-orm-panache-rest/deployment/src/main/java/io/quarkus/hibernate/orm/panache/rest/deployment/methods/MethodImplementor.java

```diff
@@ -0,0 +1,109 @@
+package io.quarkus.hibernate.orm.panache.rest.deployment.methods;
+
+import java.util.List;
+
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Type;
+import org.jboss.logging.Logger;
+
+import io.quarkus.gizmo.BytecodeCreator;
+import io.quarkus.gizmo.ClassCreator;
+import io.quarkus.gizmo.MethodCreator;
+import io.quarkus.gizmo.MethodDescriptor;
+import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
+import io.quarkus.hibernate.orm.panache.rest.PanacheCrudController;
+import io.quarkus.hibernate.orm.panache.rest.deployment.utils.PanacheRestResourceReader;
+import io.quarkus.hibernate.orm.panache.rest.runtime.hal.HalEntityWrapper;
+
+public abstract class MethodImplementor {
+
+    private static final DotName PANACHE_CRUD_CONTROLLER_INTERFACE = DotName
+            .createSimple(PanacheCrudController.class.getName());
+
+    private static final Logger LOGGER = Logger.getLogger(MethodImplementor.class);
+
+    private final ClassInfo controllerInterface;
+
+    private final PanacheRestResourceReader panacheRestResourceReader;
+
+    private final String entityType;
+
+    private final String idType;
+
+    protected MethodImplementor(ClassInfo controllerInterface, IndexView index) {
+        this.controllerInterface = controllerInterface;
+        this.panacheRestResourceReader = new PanacheRestResourceReader(index);
+
+        List<Type> types = getControllerTypes(controllerInterface);
+        this.entityType = types.get(0).name().toString();
+        this.idType = types.get(1).name().toString();
+    }
+
+    public void implement(ClassCreator classCreator) {
+        if (panacheRestResourceReader.isExposed(controllerInterface, getName(), getParameterTypes())) {
+            LOGGER.tracef("Implementing '%s#%s' method", classCreator.getClassName(), getName());
+            implementRegular(classCreator);
+            if (panacheRestResourceReader.isHal(controllerInterface, getName(), getParameterTypes())) {
+                LOGGER.tracef("Implementing '%s#%s' HAL method", classCreator.getClassName(), getName());
+                implementHal(classCreator);
+            }
+        } else {
+            LOGGER.tracef("Skipping '%s#%s', method not exposed", classCreator.getClassName(), getName());
+            implementNotExposed(classCreator);
+        }
+    }
+
+    protected abstract void implementRegular(ClassCreator classCreator);
+
+    protected abstract void implementHal(ClassCreator classCreator);
+
+    protected abstract String getName();
+
+    protected abstract String getReturnType();
```

Comment:
Minor comment: Seems like all implementations return `Response`. Can't we get rid of this?